File: sdks/community/dart/test/types/capabilities_test.dart
import 'package:ag_ui/ag_ui.dart';
import 'package:test/test.dart';

Matcher _fieldError(String field) => isA<AGUIValidationError>()
    .having((error) => error.field, 'field', field)
    .having((error) => error.json, 'json', isNull)
    .having((error) => error.cause, 'cause', isNull);

Map<String, dynamic> _completeCapabilities() => {
      'identity': {
        'name': 'Fixture Agent',
        'type': 'test',
        'description': 'synthetic',
        'version': '1.0.0',
        'provider': 'example',
        'documentationUrl': 'https://example.invalid/docs',
        'metadata': {
          'nested': {'nullable': null},
          'enabled': false,
        },
      },
      'transport': {
        'streaming': true,
        'websocket': false,
        'httpBinary': true,
        'pushNotifications': false,
        'resumable': true,
      },
      'tools': {
        'supported': true,
        'items': [
          {
            'name': 'lookup',
            'description': 'synthetic',
            'parameters': {'type': 'object', 'required': <dynamic>[]},
            'metadata': {'nested': null},
          },
        ],
        'parallelCalls': false,
        'clientProvided': true,
      },
      'output': {
        'structuredOutput': false,
        'supportedMimeTypes': ['text/plain', 'application/json'],
      },
      'state': {
        'snapshots': true,
        'deltas': true,
        'memory': false,
        'persistentState': false,
      },
      'multiAgent': {
        'supported': true,
        'delegation': true,
        'handoffs': false,
        'subAgents': [
          {'name': 'researcher', 'description': 'synthetic'},
        ],
      },
      'reasoning': {'supported': true, 'streaming': false, 'encrypted': true},
      'multimodal': {
        'input': {
          'image': true,
          'audio': false,
          'video': true,
          'pdf': false,
          'file': true,
        },
        'output': {'image': false, 'audio': true},
      },
      'execution': {
        'codeExecution': false,
        'sandboxed': true,
        'maxIterations': 1.0,
        'maxExecutionTime': 1000.0,
      },
      'humanInTheLoop': {
        'supported': true,
        'approvals': false,
        'interventions': true,
        'feedback': false,
        'interrupts': true,
        'approveWithEdits': false,
      },
      'custom': {
        'flag': false,
        'limit': 0,
        'empty': <dynamic>[],
        'nested': {'nullable': null},
      },
    };

void main() {
  group('AgentCapabilities', () {
    test('public model round-trips every shared field', () {
      final input = _completeCapabilities();
      final encoded = AgentCapabilities.fromJson(input).toJson();

      expect(encoded, input);
      expect((encoded['execution'] as Map)['maxIterations'], 1);
      expect((encoded['identity'] as Map)['metadata'], {
        'nested': {'nullable': null},
        'enabled': false,
      });
      expect(
        ((encoded['tools'] as Map)['items'] as List).single,
        containsPair('metadata', {'nested': null}),
      );
    });

    test(
      'empty and explicit false, zero, lists, maps, and objects survive',
      () {
        expect(const AgentCapabilities().toJson(), isEmpty);
        final value = AgentCapabilities.fromJson({
          'identity': <String, dynamic>{},
          'transport': {'streaming': false},
          'tools': {'supported': false, 'items': <dynamic>[]},
          'output': {'supportedMimeTypes': <dynamic>[]},
          'multiAgent': {'subAgents': <dynamic>[]},
          'execution': {'maxIterations': 0, 'maxExecutionTime': 0},
          'custom': <String, dynamic>{},
        });
        expect(value.toJson(), {
          'identity': <String, dynamic>{},
          'transport': {'streaming': false},
          'tools': {'supported': false, 'items': <dynamic>[]},
          'output': {'supportedMimeTypes': <dynamic>[]},
          'multiAgent': {'subAgents': <dynamic>[]},
          'execution': {'maxIterations': 0, 'maxExecutionTime': 0},
          'custom': <String, dynamic>{},
        });
      },
    );

    test('reads every snake_case alias and emits camelCase', () {
      final value = AgentCapabilities.fromJson({
        'identity': {'documentation_url': 'docs'},
        'transport': {'http_binary': false, 'push_notifications': true},
        'tools': {'parallel_calls': false, 'client_provided': true},
        'output': {
          'structured_output': true,
          'supported_mime_types': ['text/plain'],
        },
        'state': {'persistent_state': false},
        'multi_agent': {
          'sub_agents': [
            {'name': 'worker'},
          ],
        },
        'execution': {
          'code_execution': true,
          'max_iterations': 2.0,
          'max_execution_time': 3,
        },
        'human_in_the_loop': {'approve_with_edits': false},
      });

      expect(value.toJson(), {
        'identity': {'documentationUrl': 'docs'},
        'transport': {'httpBinary': false, 'pushNotifications': true},
        'tools': {'parallelCalls': false, 'clientProvided': true},
        'output': {
          'structuredOutput': true,
          'supportedMimeTypes': ['text/plain'],
        },
        'state': {'persistentState': false},
        'multiAgent': {
          'subAgents': [
            {'name': 'worker'},
          ],
        },
        'execution': {
          'codeExecution': true,
          'maxIterations': 2,
          'maxExecutionTime': 3,
        },
        'humanInTheLoop': {'approveWithEdits': false},
      });
    });

    test('camelCase presence wins over snake_case, including null', () {
      final value = AgentCapabilities.fromJson({
        'multiAgent': null,
        'multi_agent': {'supported': true},
        'transport': {'httpBinary': null, 'http_binary': true},
      });
      expect(value.multiAgent, isNull);
      expect(value.transport?.httpBinary, isNull);
    });

    test('nested wrong types identify complete canonical paths', () {
      final invalid = <Map<String, dynamic>, String>{
        {
          'transport': {'httpBinary': 'yes'},
        }: 'transport.httpBinary',
        {
          'multimodal': {
            'input': {'video': 1},
          },
        }: 'multimodal.input.video',
        {
          'multiAgent': {
            'subAgents': [
              {'name': false},
            ],
          },
        }: 'multiAgent.subAgents[0].name',
        {
          'tools': {
            'items': [
              {'name': 'x', 'description': false},
            ],
          },
        }: 'tools.items[0].description',
        {
          'output': {
            'supportedMimeTypes': ['text/plain', false],
          },
        }: 'output.supportedMimeTypes[1]',
        {'custom': <dynamic>[]}: 'custom',
      };
      for (final entry in invalid.entries) {
        expect(
          () => AgentCapabilities.fromJson(entry.key),
          throwsA(_fieldError(entry.value)),
          reason: entry.key.toString(),
        );
      }
    });

    test('required sub-agent name is enforced', () {
      expect(
        () => AgentCapabilities.fromJson({
          'multiAgent': {
            'subAgents': [<String, dynamic>{}],
          },
        }),
        throwsA(_fieldError('multiAgent.subAgents[0].name')),
      );
    });

    test('copy keeps omitted values and clears nullable fields', () {
      final value = AgentCapabilities.fromJson(_completeCapabilities());
      expect(value.copyWith().toJson(), value.toJson());
      expect(value.copyWith(identity: null).identity, isNull);
      expect(value.identity!.copyWith(metadata: null).metadata, isNull);
      expect(value.transport!.copyWith(httpBinary: null).httpBinary, isNull);
      expect(value.tools!.copyWith(items: null).items, isNull);
      expect(
        value.output!.copyWith(supportedMimeTypes: null).supportedMimeTypes,
        isNull,
      );
      expect(
        value.state!.copyWith(persistentState: null).persistentState,
        isNull,
      );
      expect(value.multiAgent!.copyWith(subAgents: null).subAgents, isNull);
      expect(value.reasoning!.copyWith(encrypted: null).encrypted, isNull);
      expect(value.multimodal!.copyWith(input: null).input, isNull);
      expect(
        value.execution!.copyWith(maxIterations: null).maxIterations,
        isNull,
      );
      expect(
        value.humanInTheLoop!.copyWith(approveWithEdits: null).approveWithEdits,
        isNull,
      );
      expect(value.copyWith(custom: null).custom, isNull);
    });

    test('execution limits use the manifest shared integral int64 domain', () {
      for (final value in <num>[-1, 0, 1.0]) {
        expect(
          ExecutionCapabilities.fromJson({
            'maxIterations': value,
          }).maxIterations,
          value.toInt(),
        );
      }

      for (final value in <Object?>[
        1.5,
        double.nan,
        double.infinity,
        '1',
        true,
        9.223372036854776e18,
        -9.223372036854778e18,
      ]) {
        expect(
          () => ExecutionCapabilities.fromJson({'maxExecutionTime': value}),
          throwsA(_fieldError('maxExecutionTime')),
          reason: value.toString(),
        );
      }
      expect(
        () => ExecutionCapabilities(maxIterations: 1.5),
        throwsA(_fieldError('maxIterations')),
      );
      expect(
        () => ExecutionCapabilities(maxIterations: 1)
            .copyWith(maxIterations: '1'),
        throwsA(_fieldError('maxIterations')),
      );
    });

    test(
      'execution limits accept signed int64 endpoints on the VM',
      () {
        final minimum = int.parse('-9223372036854775808');
        final maximum = int.parse('9223372036854775807');
        for (final value in <int>[minimum, maximum]) {
          expect(
            ExecutionCapabilities.fromJson({
              'maxIterations': value,
            }).maxIterations,
            value,
          );
        }
      },
      testOn: 'vm',
    );

    test(
      'unknown fields are dropped without changing arbitrary JSON values',
      () {
        final custom = {
          'nullable': null,
          'nested': [false, 0, <String, dynamic>{}],
        };
        final value = AgentCapabilities.fromJson({
          'custom': custom,
          'unknown': {'drop': true},
        });
        expect(value.toJson(), {'custom': custom});
      },
    );

    test('maps reject non-string keys with structured field paths', () {
      final invalid = <Map<String, dynamic>, String>{
        {
          'identity': {
            'metadata': <Object?, Object?>{1: 'value'},
          },
        }: 'identity.metadata',
        {
          'custom': <Object?, Object?>{1: 'value'},
        }: 'custom',
        {
          'custom': {
            'nested': <Object?, Object?>{1: 'value'},
          },
        }: 'custom.nested',
        {
          'identity': {
            'metadata': {
              'items': [
                <Object?, Object?>{1: 'value'},
              ],
            },
          },
        }: 'identity.metadata.items[0]',
        {
          'tools': {
            'items': [
              <Object?, Object?>{1: 'value'},
            ],
          },
        }: 'tools.items[0]',
        {
          'tools': {
            'items': [
              {
                'name': 'tool',
                'description': 'tool',
                'parameters': <String, dynamic>{},
                'metadata': {
                  'nested': <Object?, Object?>{1: 'value'},
                },
              },
            ],
          },
        }: 'tools.items[0].metadata.nested',
      };
      for (final entry in invalid.entries) {
        expect(
          () => AgentCapabilities.fromJson(entry.key),
          throwsA(_fieldError(entry.value)),
        );
      }
    });

    test('direct arbitrary maps are revalidated during encoding', () {
      const capabilities = AgentCapabilities(
        custom: {
          'nested': <Object?, Object?>{1: 'value'},
        },
      );
      expect(
        capabilities.toJson,
        throwsA(_fieldError('custom.nested')),
      );
    });
  });
}

File: sdks/community/go/pkg/core/types/capabilities_test.go
package types

import (
	"encoding/json"
	"reflect"
	"strings"
	"testing"
)

func capabilityPtr[T any](v T) *T { return &v }

func TestAgentCapabilitiesAllFieldsAndWireKeys(t *testing.T) {
	falseValue, trueValue := false, true
	emptyDescription := ""
	metadata := map[string]any{"nested": map[string]any{"list": []any{"value", float64(2)}}}
	custom := map[string]any{"vendor": map[string]any{"enabled": true, "levels": []any{float64(1), "two"}}}
	tools := []Tool{{Name: "lookup", Description: "Looks up a value", Parameters: map[string]any{"type": "object"}}}
	mimeTypes := []string{"text/plain", "application/json"}
	subAgents := []SubAgentInfo{{Name: "researcher", Description: &emptyDescription}}
	zero := int64(0)

	caps := AgentCapabilities{
		Identity: &IdentityCapabilities{
			Name: capabilityPtr("Example"), Type: capabilityPtr("custom"), Description: capabilityPtr("Agent"),
			Version: capabilityPtr("1.0.0"), Provider: capabilityPtr("AG-UI"),
			DocumentationURL: capabilityPtr("https://example.test/docs"), Metadata: &metadata,
		},
		Transport:  &TransportCapabilities{Streaming: &trueValue, Websocket: &falseValue, HTTPBinary: &trueValue, PushNotifications: &falseValue, Resumable: &trueValue},
		Tools:      &ToolsCapabilities{Supported: &falseValue, Items: &tools, ParallelCalls: &trueValue, ClientProvided: &falseValue},
		Output:     &OutputCapabilities{StructuredOutput: &falseValue, SupportedMimeTypes: &mimeTypes},
		State:      &StateCapabilities{Snapshots: &trueValue, Deltas: &falseValue, Memory: &trueValue, PersistentState: &falseValue},
		MultiAgent: &MultiAgentCapabilities{Supported: &trueValue, Delegation: &falseValue, Handoffs: &trueValue, SubAgents: &subAgents},
		Reasoning:  &ReasoningCapabilities{Supported: &falseValue, Streaming: &trueValue, Encrypted: &falseValue},
		Multimodal: &MultimodalCapabilities{
			Input:  &MultimodalInputCapabilities{Image: &trueValue, Audio: &falseValue, Video: &trueValue, PDF: &falseValue, File: &trueValue},
			Output: &MultimodalOutputCapabilities{Image: &falseValue, Audio: &trueValue},
		},
		Execution:      &ExecutionCapabilities{CodeExecution: &trueValue, Sandboxed: &falseValue, MaxIterations: &zero, MaxExecutionTime: capabilityPtr(int64(5000))},
		HumanInTheLoop: &HumanInTheLoopCapabilities{Supported: &trueValue, Approvals: &falseValue, Interventions: &trueValue, Feedback: &falseValue, Interrupts: &trueValue, ApproveWithEdits: &falseValue},
		Custom:         &custom,
	}

	data, err := json.Marshal(caps)
	if err != nil {
		t.Fatal(err)
	}
	var wire map[string]any
	if err := json.Unmarshal(data, &wire); err != nil {
		t.Fatal(err)
	}
	wantTop := []string{"identity", "transport", "tools", "output", "state", "multiAgent", "reasoning", "multimodal", "execution", "humanInTheLoop", "custom"}
	for _, key := range wantTop {
		if _, ok := wire[key]; !ok {
			t.Errorf("missing top-level wire key %q in %s", key, data)
		}
	}
	for _, key := range []string{"documentationUrl", "httpBinary", "pushNotifications", "parallelCalls", "clientProvided", "structuredOutput", "supportedMimeTypes", "persistentState", "subAgents", "codeExecution", "maxIterations", "maxExecutionTime", "approveWithEdits"} {
		if !strings.Contains(string(data), `"`+key+`"`) {
			t.Errorf("missing camelCase wire key %q in %s", key, data)
		}
	}

	var roundTrip AgentCapabilities
	if err := json.Unmarshal(data, &roundTrip); err != nil {
		t.Fatal(err)
	}
	if !reflect.DeepEqual(caps, roundTrip) {
		t.Errorf("round trip mismatch\nwant: %#v\n got: %#v", caps, roundTrip)
	}
}

func TestAgentCapabilitiesOptionalWireStates(t *testing.T) {
	data, err := json.Marshal(AgentCapabilities{})
	if err != nil {
		t.Fatal(err)
	}
	if string(data) != "{}" {
		t.Fatalf("empty declaration = %s, want {}", data)
	}

	emptyTools := []Tool{}
	emptyStrings := []string{}
	emptyAgents := []SubAgentInfo{}
	emptyMap := map[string]any{}
	falseValue := false
	zero := int64(0)
	caps := AgentCapabilities{
		Transport:  &TransportCapabilities{Streaming: &falseValue},
		Tools:      &ToolsCapabilities{Items: &emptyTools},
		Output:     &OutputCapabilities{SupportedMimeTypes: &emptyStrings},
		MultiAgent: &MultiAgentCapabilities{SubAgents: &emptyAgents},
		Execution:  &ExecutionCapabilities{MaxIterations: &zero},
		Custom:     &emptyMap,
	}
	data, err = json.Marshal(caps)
	if err != nil {
		t.Fatal(err)
	}
	want := `{"transport":{"streaming":false},"tools":{"items":[]},"output":{"supportedMimeTypes":[]},"multiAgent":{"subAgents":[]},"execution":{"maxIterations":0},"custom":{}}`
	if string(data) != want {
		t.Fatalf("explicit zero values = %s, want %s", data, want)
	}
}

func TestAgentCapabilitiesPermissiveDecodeAndArbitraryJSON(t *testing.T) {
	payload := `{"identity":{"metadata":{"array":[1,{"deep":null}]}},"custom":{"flag":false,"object":{"n":3}},"futureCategory":{"enabled":true}}`
	var caps AgentCapabilities
	if err := json.Unmarshal([]byte(payload), &caps); err != nil {
		t.Fatal(err)
	}
	if caps.Identity == nil || caps.Identity.Metadata == nil || caps.Custom == nil {
		t.Fatal("arbitrary maps were not decoded")
	}
	if got := (*caps.Identity.Metadata)["array"].([]any)[1].(map[string]any)["deep"]; got != nil {
		t.Fatalf("nested value = %#v, want nil", got)
	}
	if got := (*caps.Custom)["flag"]; got != false {
		t.Fatalf("custom flag = %#v, want false", got)
	}

	data, err := json.Marshal(caps)
	if err != nil {
		t.Fatal(err)
	}
	if strings.Contains(string(data), "futureCategory") {
		t.Fatalf("unknown field unexpectedly retained: %s", data)
	}
}

func TestExecutionCapabilitiesAcceptExactIntegralNumberSpellings(t *testing.T) {
	tests := []struct {
		payload string
		field   func(ExecutionCapabilities) *int64
		want    int64
	}{
		{`{"maxIterations":1.0}`, func(c ExecutionCapabilities) *int64 { return c.MaxIterations }, 1},
		{`{"maxIterations":1e3}`, func(c ExecutionCapabilities) *int64 { return c.MaxIterations }, 1000},
		{`{"maxIterations":1000e-3}`, func(c ExecutionCapabilities) *int64 { return c.MaxIterations }, 1},
		{`{"maxExecutionTime":0.000e999999999}`, func(c ExecutionCapabilities) *int64 { return c.MaxExecutionTime }, 0},
		{`{"maxExecutionTime":9223372036854775807.0}`, func(c ExecutionCapabilities) *int64 { return c.MaxExecutionTime }, 9223372036854775807},
		{`{"maxExecutionTime":-9223372036854775808e0}`, func(c ExecutionCapabilities) *int64 { return c.MaxExecutionTime }, -9223372036854775808},
	}
	for _, tt := range tests {
		var caps ExecutionCapabilities
		if err := json.Unmarshal([]byte(tt.payload), &caps); err != nil {
			t.Errorf("json.Unmarshal(%s): %v", tt.payload, err)
			continue
		}
		if got := tt.field(caps); got == nil || *got != tt.want {
			t.Errorf("json.Unmarshal(%s) = %v, want %d", tt.payload, got, tt.want)
		}
	}
}

func TestExecutionCapabilitiesRejectInvalidLimits(t *testing.T) {
	for _, payload := range []string{
		`{"maxIterations":1.5}`,
		`{"maxExecutionTime":2.25}`,
		`{"maxIterations":9007199254740991.0000000001}`,
		`{"maxIterations":9223372036854775808}`,
		`{"maxExecutionTime":-9223372036854775809.0}`,
		`{"maxIterations":1e999999999}`,
		`{"maxExecutionTime":1e-999999999}`,
		`{"maxIterations":"1"}`,
		`{"maxExecutionTime":true}`,
	} {
		var caps ExecutionCapabilities
		if err := json.Unmarshal([]byte(payload), &caps); err == nil {
			t.Errorf("json.Unmarshal(%s) succeeded, want error", payload)
		}
	}
}

func TestExecutionCapabilitiesCustomDecodeRemainsPermissive(t *testing.T) {
	var caps ExecutionCapabilities
	if err := json.Unmarshal([]byte(`{"codeExecution":false,"sandboxed":true,"maxIterations":null,"futureLimit":12}`), &caps); err != nil {
		t.Fatal(err)
	}
	if caps.CodeExecution == nil || *caps.CodeExecution || caps.Sandboxed == nil || !*caps.Sandboxed {
		t.Fatalf("boolean fields were not decoded: %#v", caps)
	}
	if caps.MaxIterations != nil {
		t.Fatalf("null maxIterations = %v, want nil", caps.MaxIterations)
	}
}

func TestExecutionCapabilitiesExponentBoundDependsOnMantissa(t *testing.T) {
	longInteger := "1" + strings.Repeat("0", 1000)
	var caps ExecutionCapabilities
	if err := json.Unmarshal([]byte(`{"maxIterations":`+longInteger+`e-1000}`), &caps); err != nil {
		t.Fatalf("exact long-mantissa integer: %v", err)
	}
	if caps.MaxIterations == nil || *caps.MaxIterations != 1 {
		t.Fatalf("long-mantissa integer = %v, want 1", caps.MaxIterations)
	}

	if err := json.Unmarshal([]byte(`{"maxIterations":`+longInteger+`e-999999999}`), &caps); err == nil {
		t.Fatal("huge negative exponent on nonzero long mantissa succeeded, want fractional error")
	}
}

func TestExecutionCapabilitiesDecodeReusesReceiverLikeEncodingJSON(t *testing.T) {
	trueValue := true
	iterations, executionTime := int64(4), int64(500)
	caps := ExecutionCapabilities{
		CodeExecution:    &trueValue,
		MaxIterations:    &iterations,
		MaxExecutionTime: &executionTime,
	}
	if err := json.Unmarshal([]byte(`{"codeExecution":false,"maxIterations":null,"future":true}`), &caps); err != nil {
		t.Fatal(err)
	}
	if caps.CodeExecution == nil || *caps.CodeExecution {
		t.Fatalf("codeExecution = %v, want explicit false", caps.CodeExecution)
	}
	if caps.MaxIterations != nil {
		t.Fatalf("maxIterations = %v, want nil after null", caps.MaxIterations)
	}
	if caps.MaxExecutionTime == nil || *caps.MaxExecutionTime != 500 {
		t.Fatalf("omitted maxExecutionTime = %v, want preserved 500", caps.MaxExecutionTime)
	}

	before := caps
	if err := json.Unmarshal([]byte(`{"sandboxed":true,"maxExecutionTime":1.5}`), &caps); err == nil {
		t.Fatal("invalid limit succeeded")
	}
	if !reflect.DeepEqual(caps, before) {
		t.Fatalf("receiver changed after failed decode\nbefore: %#v\n after: %#v", before, caps)
	}
}
